src/NHibernate/Id/SelectGenerator.cs

```diff
@@ -72,19 +81,42 @@ public partial class SelectGeneratorDelegate : AbstractSelectingDelegate
 			private readonly SqlString idSelectString;
 			private readonly IType idType;
 			private readonly IPostInsertIdentityPersister persister;
+			private readonly IEntityPersister entityPersister;
 
-			private readonly string uniqueKeyPropertyName;
-			private readonly IType uniqueKeyType;
+			private readonly string[] uniqueKeyPropertyNames;
+			private readonly IType[] uniqueKeyTypes;
 
 			internal SelectGeneratorDelegate(IPostInsertIdentityPersister persister, ISessionFactoryImplementor factory,
-			                                 string suppliedUniqueKeyPropertyName) : base(persister)
+			                                 string suppliedUniqueKeyPropertyNames) : base(persister)
 			{
 				this.persister = persister;
 				this.factory = factory;
-				uniqueKeyPropertyName = DetermineNameOfPropertyToUse((IEntityPersister) persister, suppliedUniqueKeyPropertyName);
 
-				idSelectString = persister.GetSelectByUniqueKeyString(uniqueKeyPropertyName);
-				uniqueKeyType = ((IEntityPersister) persister).GetPropertyType(uniqueKeyPropertyName);
+				if (persister is IEntityPersister entityPersister)
+				{
+					this.entityPersister = entityPersister;
+					uniqueKeyPropertyNames = DetermineNameOfPropertiesToUse(entityPersister, suppliedUniqueKeyPropertyNames);
+
+					uniqueKeyTypes = uniqueKeyPropertyNames.Select(p => entityPersister.GetPropertyType(p)).ToArray();
+				}
+				else if (persister is IQueryableCollection collectionPersister)
+				{
+					uniqueKeyPropertyNames = new string[2];
+					uniqueKeyTypes = new IType[2];
+
+					uniqueKeyPropertyNames[0] = collectionPersister.KeyColumnNames[0];
+					uniqueKeyTypes[0] = collectionPersister.KeyType;
+
+					uniqueKeyPropertyNames[1] = collectionPersister.ElementColumnNames[0];
+					uniqueKeyTypes[1] = collectionPersister.ElementType;
+				}
+				else 
+				{
+					throw new IdentifierGenerationException(
+						$"Persister type {persister.GetType()} is not supported by Select Generator");
+				}
+
+				idSelectString = persister.GetSelectByUniqueKeyString(uniqueKeyPropertyNames);
```

Comment:
After a closer look, I think this all shall be hidden inside `GetSelectByUniqueKeyString`